sycl/include/CL/sycl/detail/clusm.hpp

```diff
@@ -0,0 +1,121 @@
+//==---------------- clusm.hpp - SYCL USM for CL Utils ---------*- C++ -*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+
+#pragma once
+
+#include <map>
+#include <vector>
+#include <mutex>
+
+#include <CL/cl.h>
+#include <CL/cl_usm_ext.h>
+
+namespace cl {
+namespace sycl {
+namespace detail {
+namespace usm {
+        
+class CLUSM {
+public:
+  static bool Create(CLUSM *&pCLUSM);
+  static void Delete(CLUSM *&pCLUSM);
+
+  void initExtensions(cl_platform_id platform);
+  
+  void *hostMemAlloc(cl_context context, cl_mem_properties_intel* properties,
+                     size_t size, cl_uint alignment, cl_int *errcode_ret);
+  void *deviceMemAlloc(cl_context context, cl_device_id device,
+                       cl_mem_properties_intel* properties, size_t size,
+                       cl_uint alignment, cl_int *errcode_ret);
+  void *sharedMemAlloc(cl_context context, cl_device_id device,
+                       cl_mem_properties_intel* properties, size_t size,
+                       cl_uint alignment, cl_int *errcode_ret);
+
+  cl_int memFree(cl_context context, const void *ptr);
+
+  cl_int getMemAllocInfoINTEL(cl_context context, const void *ptr,
+                              cl_mem_info_intel param_name,
+                              size_t param_value_size, void *param_value,
+                              size_t *param_value_size_ret);
+
+  cl_int setKernelExecInfo(cl_kernel kernel, cl_kernel_exec_info param_name,
+                           size_t param_value_size, const void *param_value);
+
+  cl_int setKernelIndirectUSMExecInfo(cl_command_queue queue, cl_kernel kernel);
+
+  template <class T>
+  cl_int writeParamToMemory(size_t param_value_size, T param,
+                            size_t *param_value_size_ret, T *pointer) const;
+
+private:
+  std::mutex mLock;
+
+  CLUSM() {}
+  ~CLUSM() {}
+
+  struct SUSMAllocInfo {
+    SUSMAllocInfo()
+        : Type(CL_MEM_TYPE_UNKNOWN_INTEL), BaseAddress(NULL), Size(0),
+          Alignment(0) {}
+
+    cl_unified_shared_memory_type_intel Type;
+
+    const void *BaseAddress;
+    size_t Size;
+    size_t Alignment;
+  };
+
+  typedef std::map<const void *, SUSMAllocInfo> CUSMAllocMap;
```

Comment:
This file looks like real C++, so what about `using CUSMAllocMap = ` and so on across this file.